contracts/0.8.9/oracle/ValidatorsExitBus.sol

```diff
@@ -22,81 +27,182 @@ contract ValidatorsExitBus is AccessControlEnumerable {
+    /// @notice An ACL role granting the permission to pause accepting validator exit requests
+    bytes32 public constant PAUSE_ROLE = keccak256("PAUSE_ROLE");
+    /// @notice An ACL role granting the permission to resume accepting validator exit requests
+    bytes32 public constant RESUME_ROLE = keccak256("RESUME_ROLE");
+
+    bytes32 public constant EXIT_REQUEST_LIMIT_POSITION = keccak256("lido.ValidatorsExitBus.maxExitRequestsLimit");
+
     /// Length in bytes of packed request
     uint256 internal constant PACKED_REQUEST_LENGTH = 64;
 
     uint256 internal constant PUBLIC_KEY_LENGTH = 48;
 
+    ILidoLocator internal immutable LOCATOR;
+
+    /// @notice The list format of the validator exit requests data. Used when all
+    /// requests fit into a single transaction.
+    ///
+    /// Each validator exit request is described by the following 64-byte array:
+    ///
+    /// MSB <------------------------------------------------------- LSB
+    /// |  3 bytes   |  5 bytes   |     8 bytes      |    48 bytes     |
+    /// |  moduleId  |  nodeOpId  |  validatorIndex  | validatorPubkey |
+    ///
+    /// All requests are tightly packed into a byte array where requests follow
+    /// one another without any separator or padding, and passed to the `data`
+    /// field of the report structure.
+    ///
+    /// Requests must be sorted in the ascending order by the following compound
+    /// key: (moduleId, nodeOpId, validatorIndex).
+    ///
+    uint256 public constant DATA_FORMAT_LIST = 1;
+
     /// Hash constant for mapping exit requests storage
-    bytes32 internal constant EXIT_REQUESTS_HASHES_POSITION =
-        keccak256("lido.ValidatorsExitBus.reportHashes");
+    bytes32 internal constant EXIT_REQUESTS_HASHES_POSITION = keccak256("lido.ValidatorsExitBus.reportHashes");
+
+    /// @dev Ensures the contract’s ETH balance is unchanged.
+    modifier preservesEthBalance() {
+        uint256 balanceBeforeCall = address(this).balance - msg.value;
+        _;
+        assert(address(this).balance == balanceBeforeCall);
+    }
+
+    constructor(address lidoLocator) {
+        LOCATOR = ILidoLocator(lidoLocator);
+    }
+
+    function submitReportHash(bytes32 exitReportHash) external whenResumed onlyRole(SUBMIT_REPORT_HASH_ROLE) {
+        uint256 contractVersion = getContractVersion();
+        _storeExitRequestHash(exitReportHash, type(uint256).max, 0, contractVersion, DeliveryHistory(0, 0));
+    }
+
+    function emitExitEvents(ExitRequestData calldata request, uint256 contractVersion) external whenResumed {
+        bytes calldata data = request.data;
+        _checkContractVersion(contractVersion);
+
+        RequestStatus storage requestStatus = _storageExitRequestsHashes()[
+            keccak256(abi.encode(data, request.dataFormat))
+        ];
+
+        if (requestStatus.contractVersion == 0) {
+            revert ExitHashWasNotSubmitted();
+        }
+
+        if (request.dataFormat != DATA_FORMAT_LIST) {
```

Comment:
in the context of the comment about `contractVersion` above, is there a case when the `DATA_FORMAT_LIST` constant is changed without changing the contract version? i.e. it seems that the simultaneous presence of both checks is a bit redundant